src/which.js

```diff
@@ -45,74 +39,60 @@ function _which(options, cmd) {
 
   var pathEnv = process.env.path || process.env.Path || process.env.PATH;
   var pathArray = splitPath(pathEnv);
-  var all = options.all;
-  var where = all ? [] : null;
+
+  var queryMatches = [];
 
   // No relative/absolute paths provided?
-  if (cmd.search(/\//) === -1) {
+  if (cmd.indexOf('/') === -1) {
+    // Assume that there are no extensions to append to queries (this is the
+    // case for unix)
+    var pathExtArray = [''];
+    if (common.platform === 'win') {
+      // In case the PATHEXT variable is somehow not set (e.g.
+      // child_process.spawn with an empty environment), use the XP default.
+      var pathExtEnv = process.env.PATHEXT || XP_DEFAULT_PATHEXT;
+      pathExtArray = splitPath(pathExtEnv.toUpperCase());
+    }
+
     // Search for command in PATH
-    pathArray.forEach(function (dir) {
-      if (!all && where) return; // already found it
+    for (var k = 0; k < pathArray.length; k++) {
+      // already found it
+      if (queryMatches.length > 0 && !options.all) break;
 
-      var attempt = path.resolve(dir, cmd);
+      var attempt = path.resolve(pathArray[k], cmd);
 
       if (common.platform === 'win') {
         attempt = attempt.toUpperCase();
+      }
 
-        // In case the PATHEXT variable is somehow not set (e.g.
-        // child_process.spawn with an empty environment), use the XP default.
-        var pathExtEnv = process.env.PATHEXT || XP_DEFAULT_PATHEXT;
-        var pathExtArray = splitPath(pathExtEnv.toUpperCase());
-        var i;
-
-        // If the extension is already in PATHEXT, just return that.
-        for (i = 0; i < pathExtArray.length; i++) {
-          var ext = pathExtArray[i];
-          if (attempt.slice(-ext.length) === ext && checkPath(attempt)) {
-            if (all) {
-              where.push(attempt);
-            } else {
-              where = attempt;
-              return;
-            }
-          }
-        }
-
-        // Cycle through the PATHEXT variable
-        var baseAttempt = attempt;
-        for (i = 0; i < pathExtArray.length; i++) {
-          attempt = baseAttempt + pathExtArray[i];
-          if (checkPath(attempt)) {
-            if (all) {
-              where.push(attempt);
-            } else {
-              where = attempt;
-              return;
-            }
-          }
-        }
-      } else {
-        // Assume it's Unix-like
+      var match = attempt.match(/\.[^<>:"/\|?*.]+$/);
+      if (match && pathExtArray.search(match[0]) >= 0) { // this is Windows-only
```

Comment:
Ah, this was a mistake. `Array.prototype.search` isn't even defined... `indexOf` is what I was looking for. Thanks